mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdater.java

```diff
@@ -75,63 +79,65 @@ public boolean isHandling(Event event) {
-                        .quotaSize(quotaManager.getStorageQuota(quotaRoot))
-                        .instant(Instant.now())
-                        .build(),
-                    NO_REGISTRATION_KEYS)).sneakyThrow()))
-                .block();
-        }).sneakyThrow());
+    private Mono<Void> handleAddedEvent(Added added, QuotaRoot quotaRoot) {
+        return computeQuotaOperation(added, quotaRoot)
+            .flatMap(quotaOperation ->
+                Mono.from(currentQuotaManager.increase(quotaOperation))
+                    .then(dispatchNewQuota(quotaRoot, added.getUsername())));
     }
 
-    private void handleAddedEvent(Added added, QuotaRoot quotaRoot) {
-        computeQuotaOperation(added, quotaRoot).ifPresent(Throwing.<QuotaOperation>consumer(quotaOperation -> {
-            Mono.from(currentQuotaManager.increase(quotaOperation))
-                .then(Mono.defer(Throwing.supplier(() -> eventBus.dispatch(
-                    EventFactory.quotaUpdated()
-                        .randomEventId()
-                        .user(added.getUsername())
-                        .quotaRoot(quotaRoot)
-                        .quotaCount(quotaManager.getMessageQuota(quotaRoot))
-                        .quotaSize(quotaManager.getStorageQuota(quotaRoot))
-                        .instant(Instant.now())
-                        .build(),
-                    NO_REGISTRATION_KEYS)).sneakyThrow()))
-                .block();
-        }).sneakyThrow());
+    private Mono<Void> dispatchNewQuota(QuotaRoot quotaRoot, Username username) {
+        Mono<Quota<QuotaCountLimit, QuotaCountUsage>> messageQuota = Mono.fromCallable(() -> quotaManager.getMessageQuota(quotaRoot));
+        Mono<Quota<QuotaSizeLimit, QuotaSizeUsage>> storageQuota = Mono.fromCallable(() -> quotaManager.getStorageQuota(quotaRoot));
+
+        Mono<Tuple2<Quota<QuotaCountLimit, QuotaCountUsage>, Quota<QuotaSizeLimit, QuotaSizeUsage>>> quotasMono =
+            messageQuota.zipWith(storageQuota)
+                .subscribeOn(Schedulers.elastic());
+
+        return quotasMono
+            .flatMap(quotas -> eventBus.dispatch(
+                EventFactory.quotaUpdated()
+                    .randomEventId()
+                    .user(username)
+                    .quotaRoot(quotaRoot)
+                    .quotaCount(quotas.getT1())
+                    .quotaSize(quotas.getT2())
+                    .instant(Instant.now())
+                    .build(),
+                NO_REGISTRATION_KEYS));
     }
 
-    private Optional<QuotaOperation> computeQuotaOperation(MetaDataHoldingEvent metaDataHoldingEvent, QuotaRoot quotaRoot) {
+    private Mono<QuotaOperation> computeQuotaOperation(MetaDataHoldingEvent metaDataHoldingEvent, QuotaRoot quotaRoot) {
         long size = totalSize(metaDataHoldingEvent);
         long count = Integer.toUnsignedLong(metaDataHoldingEvent.getUids().size());
 
         if (count != 0 && size != 0) {
```

Comment:
I think we should extract to method?

Reply:
Sure, it would benefit from it.

I think @Arsnael did some change related to this line.

My changeset is orthogonal to this line, would you mind if I don't address it here?